### ISSUE TITLE
Unable to load live dependency into virtualenv when using existing Apache installation

### ISSUE THREAD

**Issue by undu** [GrahamDumpleton/mod_wsgi#244]
Hi,
I've been trying to load a Flask application with mod_wsgi. I've been successful on centos 6 (Apache 2.2), but I'm struggling to make it work on centos 7 (Apache 2.4).

I'm using a custom python installation, with a pip-installed mod_wsgi (latest as of right now).

On centos 6 both using an existing Apache installation and using `virtualenv/bin/mod_wsgi-express start-server /path/app.wsgi` work fine.

On centos 7, however only the `mod_wsgi-express` method works, I've tried both a 2.7.14 and 3.6.3 virtualenvs.

I've made sure both installations are as equivalent as possible, using Ansible

When using the Apache installation this error happens:

```
[[Tue Nov 21 13:20:19.030074 2017] [mpm_prefork:notice] [pid 17975] AH00163: Apache/2.4.6 (CentOS) OpenSSL/1.0.2k-fips mod_wsgi/4.5.22 Python/3.6 configured -- resuming normal operations
[Tue Nov 21 13:20:19.030101 2017] [core:notice] [pid 17975] AH00094: Command line: '/usr/sbin/httpd -D FOREGROUND'
[Tue Nov 21 13:20:43.218791 2017] [wsgi:error] [pid 17977] [remote 172.28.128.1:54412] mod_wsgi (pid=17977): Target WSGI script '/var/www/vhosts/tenochtitlan/service.wsgi' cannot be loaded as Python module.
[Tue Nov 21 13:20:43.218903 2017] [wsgi:error] [pid 17977] [remote 172.28.128.1:54412] mod_wsgi (pid=17977): Exception occurred processing WSGI script '/var/www/vhosts/tenochtitlan/service.wsgi'.
[Tue Nov 21 13:20:43.226687 2017] [wsgi:error] [pid 17977] [remote 172.28.128.1:54412] Traceback (most recent call last):
[Tue Nov 21 13:20:43.226730 2017] [wsgi:error] [pid 17977] [remote 172.28.128.1:54412]   File "/var/www/vhosts/tenochtitlan/service.wsgi", line 1, in <module>
[Tue Nov 21 13:20:43.226736 2017] [wsgi:error] [pid 17977] [remote 172.28.128.1:54412]     from cwltoolservice.cwl_flask import APP as application
[Tue Nov 21 13:20:43.226751 2017] [wsgi:error] [pid 17977] [remote 172.28.128.1:54412] ModuleNotFoundError: No module named 'cwltoolservice'
[Tue Nov 21 13:20:44.116261 2017] [wsgi:error] [pid 17977] [remote 172.28.128.1:54414] mod_wsgi (pid=17977): Target WSGI script '/var/www/vhosts/tenochtitlan/service.wsgi' cannot be loaded as Python module.
```

This dependency is my application, I install it in the virtual environment using `pip install -e .`
I've checked the virtual environment and in both installation `pip freeze` return exactly the same packages and versions.
I've replaced the `.wsgi` to some code to print installed module; cwltoolservice is completely missing from the list of modules.

For reference, this is the `.conf` that I've added to the Apache configuration.

```
LoadModule wsgi_module "/home/someone/tenochtitlan.vm/cwl-virtualenv/lib/python3.6/site-packages/mod_wsgi/server/mod_wsgi-py36.cpython-36m-x86_64-linux-gnu.so"
WSGIPythonHome "/home/someone/tenochtitlan.vm/cwl-virtualenv"

ServerName tenochtitlan
WSGISocketPrefix /var/run/wsgi

<VirtualHost *:80>
  WSGIDaemonProcess service threads=4 user=someone group=someone
  WSGIScriptAlias / /var/www/vhosts/tenochtitlan/service.wsgi
  <Directory /var/www/vhosts/tenochtitlan>
    WSGIApplicationGroup %{GLOBAL}
    WSGIProcessGroup service
    <IfVersion < 2.4>
    Order allow,deny
    Allow from all
    </IfVersion>
    <IfVersion >= 2.4>
    Require all granted
    </IfVersion>
  </Directory>
</VirtualHost>
```

I'm surely missing something here, but I do not know what is it, any ideas where the problem might be? I posted here because it might be a bug, but I do not know where, exactly.

**Comment by GrahamDumpleton** [comment]
You don't appear to be telling mod_wsgi where your application source code is.

Read:

* http://modwsgi.readthedocs.io/en/develop/user-guides/virtual-environments.html

about setting up use of Python virtual environments. At the bottom there is also a bit about adding extra directories to search for Python modules, such as where your application code is.

**Comment by undu** [comment]
I am telling mod_wsgi where the application source code is through an editable install if my application's source code inside the virtual environment, that is, installed through `pip install -e`.

Read more about editable installs [here](http://pip.readthedocs.io/en/latest/reference/pip_install/#editable-installs).

To make it more clear, that command creates a file in containing the location of the application's source code in `${WSGIPythonHome}/lib/python3.6/site-packages`.

```
$ cat ${WSGIPythonHome}/lib/python3.6/site-packages/cwltool-service.egg-info
/tmp/wes-server-master/
```

**Comment by GrahamDumpleton** [comment]
What are the permissions on the directory ``/tmp/wes-server-master/`` and what is in it? Can you show ``ls -las`` output for that directory.

**Comment by undu** [comment]
Here's the output:
```
$ ls -las /tmp/wes-server-master/
total 24
0 drwxrwxr-x.  9 someone someone  226 Nov 20 11:03 .
4 drwxrwxrwt. 16 root    root    4096 Nov 22 09:22 ..
0 drwxrwxr-x.  3 someone someone  269 Nov 20 14:46 cwltoolservice
0 drwxrwxr-x.  2 someone someone  150 Nov 20 11:03 cwltool_service.egg-info
0 drwxrwxr-x.  8 someone someone  198 Nov 21 11:25 .git
4 -rw-rw-r--.  1 someone someone  176 Nov 20 11:03 .gitignore
0 drwxrwxr-x.  2 someone someone   90 Nov 20 11:03 instance
4 -rw-rw-r--.  1 someone someone 1618 Nov 20 11:03 README.md
4 -rw-rw-r--.  1 someone someone  179 Nov 20 11:03 requirements.txt
4 -rw-rw-r--.  1 someone someone 1314 Nov 20 11:03 setup.py
0 drwxrwxr-x.  3 someone someone   19 Nov 20 11:03 swagger
0 drwxrwxr-x.  2 someone someone  105 Nov 20 11:03 testdata
0 drwxrwxr-x.  3 someone someone   67 Nov 20 11:03 tests
4 -rw-rw-r--.  1 someone someone  394 Nov 20 11:03 .travis.yml
```

**Comment by GrahamDumpleton** [comment]
What are the permissions of any files related to the package under ``${WSGIPythonHome}/lib/python3.6/site-packages/``.

**Comment by undu** [comment]
The file permissions are the same in both systems (with different Apache versions) I've tried:
```
$ ls -las tenochtitlan.vm/cwl-virtualenv/lib/python2.7/site-packages/ | grep egg-link
   4 -rw-rw-r--.   1 someone someone      24 Nov 22 09:57 cwltool-service.egg-link
```

**Comment by GrahamDumpleton** [comment]
Can only suggest you add debug to ``/var/www/vhosts/tenochtitlan/service.wsgi`` to print out value of ``sys.path`` before doing any imports to see how the differ.

**Comment by undu** [comment]
Indeed,  '/tmp/wes-server-master' is not present to sys.path

However, when I use the virtualenv's python CLI directly the directory _is_ present:

```
$ tenochtitlan.vm/cwl-virtualenv/bin/python3.6
Python 3.6.3 (default, Nov 21 2017, 10:42:31)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-16)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> print(sys.path)
['', '/home/someone/tenochtitlan.vm/cwl-virtualenv/lib/python36.zip', '/home/someone/tenochtitlan.vm/cwl-virtualenv/lib/python3.6', '/home/someone/tenochtitlan.vm/cwl-virtualenv/lib/python3.6/lib-dynload', '/home/someone/.local/lib/python3.6', '/home/someone/tenochtitlan.vm/cwl-virtualenv/lib/python3.6/site-packages', '/tmp/wes-server-master']
```

**Comment by undu** [comment]
I activated the virtual environment, as you recommend in the documentation. The issue still persists.

**Comment by GrahamDumpleton** [comment]
Did you ever sort this out? Did you ensure you had the latest ``pip`` and ``setuptools`` installed in the virtual environment? The ``pip`` and ``setuptools`` versions packaged in CentOS are old and have various issues which are fixed in newer versions.

Looping back to look at this only now as when this was raised was bad time for me as I was extremely busy through to late December. Finally catching up on lot of old mod_wsgi issues while am on holidays.

FWIW, I can't duplicate an issue with using editable ``pip`` installed modules on MacOS X using ``pip`` version 9.0.1 and ``setuptools`` version 38.4.0.

**Comment by undu** [comment]
Unfortunately I didn't.
I ended up using apache as a reverse proxy for Gunicorn.

As for `pip` and `setuptools` I was compiling python from source, executing `ensurepip`and finally using `pip`to install `setuptools`; I used this ansible recipe: https://github.com/psafont/python-from-src/blob/master/tasks/main.yml

I checked the versions in a test environment and both were on the latest version, but the environment was already set up to use gunicorn, so I don't know which versions I was using late November.
